pkg/server/api_v2_grants.go

```diff
@@ -0,0 +1,365 @@
+//
+// If the user does not have the required permissions to access the database we will return a 403 Forbidden error.
+//
+// ---
+// parameters:
+//
+//   - name: dbId
+//     type: integer
+//     description: The ID of the database to get grants for.
+//     in: path
+//
+// produces:
+// - application/json: databaseGrantsResponse
+//
+// responses:
+//
+//	"200":
+//		description: A paginated response of grants on the provided database.
+//	"404":
+//		description: The db does not exist.
+//	"400":
+//		description: The request is malformed.
+//	"500":
+//		description: An internal server error occurred.
+func (a *apiV2Server) getDatabaseGrants(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	ctx = a.sqlServer.AnnotateCtx(ctx)
+	sqlUser := authserver.UserFromHTTPAuthInfoContext(ctx)
+	if r.Method != http.MethodGet {
+		http.Error(w, "Method Not Allowed", http.StatusMethodNotAllowed)
+		return
+	}
+
+	// Validate request parameters.
+	pathVars := mux.Vars(r)
+	dbId, err := strconv.Atoi(pathVars[dbIdPathVar])
+	if err != nil || dbId < 0 {
+		http.Error(w, "invalid database id", http.StatusBadRequest)
+		return
+	}
+	queryValues := r.URL.Query()
+	pageSize, err := apiutil.GetIntQueryStringVal(queryValues, pageSizeKey)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	pageNum, err := apiutil.GetIntQueryStringVal(queryValues, pageNumKey)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	offset := 0
+	if pageNum > 0 {
+		offset = (pageNum - 1) * pageSize
+	}
+
+	// Get the database grants.
+	resp, err := getDatabaseGrantsResponseWithPagination(ctx, a.sqlServer.internalExecutor, dbId, sqlUser, pageSize, offset)
+	if err != nil {
+		srverrors.APIV2InternalError(ctx, err, w)
+		return
+	}
+
+	if resp.Name == "" {
+		http.Error(w, "database does not exist", http.StatusNotFound)
+		return
+	}
+
+	resp.PaginationInfo.PageNum = pageNum
+	resp.PaginationInfo.PageSize = pageSize
+
+	apiutil.WriteJSONResponse(ctx, w, http.StatusOK, resp)
+}
+
+func getDatabaseGrantsResponseWithPagination(
+	ctx context.Context,
+	ie *sql.InternalExecutor,
+	dbId int,
+	sqlUser username.SQLUsername,
+	limit int,
+	offset int,
+) (resp databaseGrantsResponseWithPagination, retErr error) {
+	row, err := ie.QueryRowEx(
+		ctx, "get-database-name-by-id", nil, /* txn */
+		sessiondata.NodeUserSessionDataOverride, `
+SELECT name FROM system.namespace
+WHERE "parentID" = 0 AND "parentSchemaID" = 0 AND id = $1`, dbId)
+	if err != nil {
+		return resp, err
+	}
+
+	if row == nil || row[0] == tree.DNull {
+		// Database id does not exist.
+		return resp, nil
+	}
+
+	dbName := string(*row[0].(*tree.DString))
+	resp.Name = dbName
+
+	escDbName := tree.NameStringP(&dbName)
+
+	query := safesql.NewQuery()
+	query.Append(fmt.Sprintf(`
+SELECT grantee, privilege_type, count(*) OVER() as total_row_count
+FROM [SHOW GRANTS ON DATABASE %s]`, escDbName))
```

Comment:
Is this the query that we want? I see in the figma that we are keeping the same UI where we group by userId. If we are doing this aggregation on the front end, then it could result in  an incomplete of a user's grants if they are split over 2 pages.

Reply:
Yeah I deliberated on this for a while and it seemed to me that it would be more flexible to show the grants per row rather than aggregating in the API right now, since we can extend the API to filter by users. The pagination also made more sense when we didn't aggregate the privileges by user, since the payload size could be determined by the page size given. Let's discuss this offline with Kevin and Inna on if we can swap to an unaggregated view in the UI.